asst2/CODE/my_malloc.c

```diff
@@ -4,39 +4,251 @@
+	return curr -> size & 0x7f;
+}
+
+
+//prints any error messages followed by memory 
+void error_message(char* error, char* file, int line) {
+	printf("%s:%d: %s\n", file, line, error);
+	exit(0);
+}
+
+//takes the number of segments within page to be free and writes that into metadata
+void write_free_size(metadata* curr, size_t newSize) {
+	curr -> size = newSize & 0x7f;
+	return;
+}
+
+//takes the number of segments within page to be malloc'd
+//and makes the leftmost bit of the byte-sized number 1
+//to represent malloc'd and writes this into metadata
+//if the current free block is being split with this malloc
+//also creates a metadata block to adjust the size of free userdata available
+void write_occupied_size(metadata* curr, unsigned char curr_seg, size_t newSize) {
+	//if the size of the free block is more than the new size,
+	//need to split the block into a first block which is malloc'd
+	//and second block which is still free but has a smaller size 
+	if (newSize < curr_seg) write_free_size(curr+newSize, (curr_seg - newSize));
+	curr -> size = (newSize & 0x7f) | 0x80;
+}
+
+int is_occupied_page(pagedata* curr) {
+	return  (curr->p_ind) & 0x8000;
+}
+void write_free_page(pagedata* curr, unsigned short index) {
+  curr->p_ind = (index & 0x7fff);
 }
 
-/**
- * Frees data allocated by memory management system during process teardown.
- */
-void free_mem_metadata() {
-  free(act);
-  // oops, we probably can't do this
-  free(MEM);
+void write_occupied_page(pagedata* curr, uint32_t pid, unsigned short index) {
+  curr->pid = pid;
+  curr->p_ind = (index & 0x7fff) | 0x8000;
 }
+//for debugging and error reporting purposes
+//prints memory as index of memory
+//followed by 0 for free or any other integer for occupied
+//followed by the size of user data
+void printMemory() {
+	printf("-----------------------MEMORY-----------------------\n");
+	for (int i = 0; i < num_pages; i++) {
+		pagedata* pdata = (pagedata*)myblock + i;	
+		if (!is_occupied_page(pdata)) {continue;}
+		metadata* start = (metadata*) ((char*)mem_space + i*page_size);
+		printf("--------------------PAGE--------------------\n");
+		int j = 0;
+		while (j < num_segments) {
+			metadata* mdata = start + j;
+			char curr_seg = block_size(mdata);
+			int isOcc = is_occupied(mdata);
+			printf("page[%d][%d] pid %d %p: %d  %hu %p\n", i, j, pdata->pid, mdata, isOcc, curr_seg, (start+num_segments) + j*SEGMENTSIZE);
+			j+=curr_seg;	
+		}
+	}
+	printf("-----------------------MEMORY-----------------------\n");
+}
+
+void initialize_pages() {
+	for (unsigned short i = 0; i < num_pages; i++) {
+		pagedata* pdata = (pagedata*)myblock + i;	
+		write_free_page(pdata, i);
+		metadata* mdata = (metadata*) ((char*)mem_space + i*page_size);
+		write_free_size(mdata, num_segments);
+	}
+}
+
+
+//returns to the user a pointer to the amount of memory requested
+void* myallocate(size_t size, char* file, int line, int threadreq){
+  	enter_scheduler(&timer_pause_dump);
```

Comment:
does this work if we don't have a scheduler inited?

Reply:
o lel hmm hasn't given an issue actually idk